session.go

```diff
@@ -0,0 +1,229 @@
+	b.m.Lock()
+	defer b.m.Unlock()
+
+	e := b.PushBack(v)
+
+	if b.Len() > b.size {
+		b.Remove(b.Front())
+	}
+
+	return e
+}
+
+type sessionState int
+
+const (
+	sessionStateNew sessionState = 1 << iota
+	sessionStateConnected
+	sessionStateDisconnected
+)
+
+type session struct {
+	b *buffer
+
+	conn  Conn
+	connm sync.Mutex
+
+	st  sessionState
+	stm sync.Mutex
+
+	id  uint32
+	idm sync.Mutex
+
+	err SessionError
+}
+
+func NewSession(conn Conn) Session {
+	return &session{
+		st:    sessionStateNew,
+		stm:   sync.Mutex{},
+		conn:  conn,
+		connm: sync.Mutex{},
+		idm:   sync.Mutex{},
+		b:     newBuffer(50),
+	}
+}
+
+func (s *session) Connect() error {
+	if s.isNew() {
+		return errors.New("can't connect unless the session is new")
+	}
+
+	go s.readErrors()
+	return nil
+}
+
+func (s *session) isNew() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st != sessionStateNew
+}
+
+func (s *session) Disconnected() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st == sessionStateDisconnected
+}
+
+func (s *session) Connnected() bool {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	return s.st == sessionStateConnected
+}
+
+func (s *session) Send(n Notification) error {
+	// If disconnected, error out
+	if !s.Connnected() {
+		return errors.New("not connected")
+	}
+
+	// Set identifier if not specified
+	n.Identifier = s.determineIdentifier(n.Identifier)
+
+	// Serialize
+	b, err := n.ToBinary()
+	if err != nil {
+		return err
+	}
+
+	// Add to buffer
+	s.b.Add(n)
+
+	// Send synchronously
+	return s.send(b)
+}
+
+func (s *session) send(b []byte) error {
+	s.connm.Lock()
+	defer s.connm.Unlock()
+
+	_, err := s.conn.Write(b)
+	if err == io.EOF {
+		s.Disconnect()
+		return err
+	}
+
+	return err
+}
+
+func (s *session) Disconnect() {
+	// Disconnect
+	s.transitionState(sessionStateDisconnected)
+}
+
+func (s *session) RequeueableNotifications() []Notification {
+	notifs := []Notification{}
+
+	// If still connected, return nothing
+	if s.st != sessionStateDisconnected {
+		return notifs
+	}
+
+	// Walk back to last known good notification and return the slice
+	var e *list.Element
+	for e = s.b.Front(); e != nil; e = e.Next() {
+		if n, ok := e.Value.(Notification); ok && n.Identifier == s.err.Notification.Identifier {
+			break
+		}
+	}
+
+	// Start right after errored ID and get the rest of the list
+	for e = e.Next(); e != nil; e = e.Next() {
+		n, ok := e.Value.(Notification)
+		if !ok {
+			continue
+		}
+
+		notifs = append(notifs, n)
+	}
+
+	return notifs
+}
+
+func (s *session) transitionState(st sessionState) {
+	s.stm.Lock()
+	defer s.stm.Unlock()
+
+	s.st = st
+}
+
+func (s *session) determineIdentifier(n uint32) uint32 {
+	s.idm.Lock()
+	defer s.idm.Unlock()
+
+	// If the id passed in is 0, that means it wasn't
+	// set so get the next ID. Otherwise, set it to that
+	// identifier.
+	if n == 0 {
+		s.id++
+	} else {
+		s.id = n
+	}
+
+	return s.id
+}
+
+func (s *session) readErrors() {
```

Comment:
Some details from Apple on error checking:
https://developer.apple.com/library/ios/technotes/tn2265/_index.html#//apple_ref/doc/uid/DTS40010376-CH1-TNTAG44

> If your development tool of choice supports multiple threads or interprocess communication, you could have a thread or process waiting for an error response all the time and let the main sending thread or process know when it should give up and retry.

We launch readErrors on a goroutine, but it looks like it will only read once and then end. I assume read is blocking though, but is this enough?